ign_ros2_control/src/MimicJointSystem.cc

```diff
@@ -0,0 +1,413 @@
+
+public:
+  ignition::gazebo::Model model {ignition::gazebo::kNullEntity};
+
+/// \brief Position PID controller.
+
+public:
+  ignition::math::PID posPid;
+
+/// \brief Joint index to be used.
+
+public:
+  unsigned int jointIndex = 0u;
+
+public:
+  unsigned int mimicJointIndex = 0u;
+
+public:
+  double deadZone = 1e-6;
+
+  // approach adopted from https://github.com/gazebosim/gz-sim/blob/330eaf2f135301e90096fe91897f052cdaa77013/src/systems/joint_position_controller/JointPositionController.cc#L69-L77
+/// \brief Operation modes
+  enum OperationMode
+  {
+    /// \brief Use PID to achieve positional control
+    PID,
+    /// \brief Bypass PID completely. This means the joint will move to that
+    /// position bypassing the physics engine.
+    ABS
+  };
+
+/// \brief Joint position mode
+
+public:
+  OperationMode mode = OperationMode::PID;
+
+};
+
+MimicJointSystem::MimicJointSystem()
+: dataPtr(std::make_unique < ign_ros2_control::MimicJointSystemPrivate > ())
+{
+}
+
+void MimicJointSystem::Configure(
+  const ignition::gazebo::Entity & _entity,
+  const std::shared_ptr < const sdf::Element > & _sdf,
+  ignition::gazebo::EntityComponentManager & _ecm,
+  ignition::gazebo::EventManager & /* _eventMgr*/)
+{
+
+  // Make sure the controller is attached to a valid model
+  this->dataPtr->model = ignition::gazebo::Model(_entity);
+  if (!this->dataPtr->model.Valid(_ecm)) {
+
+    ignerr << "MimicJointSystem  Failed to initialize because [" <<
+      this->dataPtr->model.Name(_ecm) <<
+      "] (Entity=" << _entity << ")] is not a model.. ";
+    return;
+  }
+
+  // Get params from SDF
+  this->dataPtr->jointName = _sdf->Get < std::string > ("joint_name");
+
+  if (this->dataPtr->jointName.empty()) {
+    ignerr << "MimicJointSystem found an empty joint_name parameter. "
+           << "Failed to initialize.";
+    return;
+  }
+
+  this->dataPtr->mimicJointName = _sdf->Get < std::string > ("mimic_joint_name");
+  if (this->dataPtr->mimicJointName.empty()) {
+    ignerr << "MimicJointSystem found an empty mimic_joint_name parameter. "
+           << "Failed to initialize.";
+    return;
+  }
+
+  if (_sdf->HasElement("multiplier")) {
+    this->dataPtr->multiplier = _sdf->Get < double > ("multiplier");
+  }
+
+  if (_sdf->HasElement("offset")) {
+    this->dataPtr->offset = _sdf->Get < double > ("offset");
+  }
+
+  if (_sdf->HasElement("joint_index")) {
+    this->dataPtr->jointIndex = _sdf->Get < unsigned int > ("joint_index");
+  }
+
+  if (_sdf->HasElement("mimic_joint_index")) {
+    this->dataPtr->mimicJointIndex = _sdf->Get < unsigned int > ("mimic_joint_index");
+  }
+
+  if (_sdf->HasElement("dead_zone")) {
+    this->dataPtr->deadZone = _sdf->Get < double > ("dead_zone");
+  }
+
+  // PID parameters
+  double p = 1;
+  double i = 0.1;
+  double d = 0.01;
+  double iMax = 1;
+  double iMin = -1;
+  double cmdMax = 1000;
+  double cmdMin = -1000;
+  double cmdOffset = 0;
+
+  if (_sdf->HasElement("p_gain")) {
```

Comment:
@kylc  This is not ros component at all and it is intended to stay like that. I made this as pure ignition plugin.

Regarding the ROS support for mimic joints I implemented similar thing in `ignition_system.cpp`. See [here](https://github.com/livanov93/ign_ros2_control/blob/1a89e0664955db392271522172c3213b2023b4c1/ign_ros2_control/src/ign_system.cpp#L975-L998) . The parameters are read every control step [here](https://github.com/livanov93/ign_ros2_control/blob/1a89e0664955db392271522172c3213b2023b4c1/ign_ros2_control/src/ign_system.cpp#L782-L783) using `generate_parameter_library`. The namespaces under they appear are the same as the name of ros2_control hardware interface --> name tag for ur robot [here](https://github.com/livanov93/Universal_Robots_ROS2_Description/blob/3a862130e5ca74e46cf403501ee2f982ff4bc1b4/urdf/ur.ros2_control.xacro#L23) and for instance of robotiq gripper name tag [here](https://github.com/livanov93/robotiq/blob/3b1e9e0f1060ca1640d838e8079e2ad74ef73350/robotiq_2f_85_gripper_visualization/urdf/robotiq_gripper.ros2_control.xacro#L7).  Using rqt_reconfigure it can be tuned quite easily.